File: p.10/Ruby/solution.rb
def prime_sieve(n)
  sieve = Array.new(n + 1, true)
  sieve[0] = sieve[1] = false
  limit = Math.sqrt(n).to_i

  (2..limit).each do |elem|
    next unless sieve[elem]

    (elem * elem..n).step(elem) do |multiple|
      sieve[multiple] = false
    end
  end

  sieve.each_index.select { |i| sieve[i] }
end

def main
  primes = prime_sieve(2_000_000)
  p primes.sum
end

main if __FILE__ == $0

File: p.10/Python/solution.py
from typing import List

def prime_sieve(n: int) -> List[int]:
    sieve = [True] * (n + 1)
    for elem in range(2, int(n ** 0.5) + 1):
        if sieve[elem]:
            for multiple in range(elem * elem, n + 1, elem):
                sieve[multiple] = False
    return [elem for elem in range(2, n) if sieve[elem]]

def main():
    print(sum(prime_sieve(2000_000)))

if __name__ == "__main__":
    main()


